x-pack/plugins/enterprise_search/public/applications/app_search/components/document_creation/document_creation_logic.ts

```diff
@@ -74,5 +99,122 @@ export const DocumentCreationLogic = kea<
+      },
+    ],
+    summary: [
+      {} as DocumentCreationSummary,
+      {
+        setSummary: (_, { summary }) => summary,
+      },
+    ],
+  }),
+  listeners: ({ values, actions }) => ({
+    onSubmitFile: async () => {
+      const { fileInput } = values;
+
+      if (!fileInput) {
+        return actions.setErrors([DOCUMENT_CREATION_ERRORS.NO_FILE]);
+      }
+      try {
+        const textInput = await readUploadedFileAsText(fileInput);
+        actions.setTextInput(textInput);
+        actions.onSubmitJson();
+      } catch {
+        actions.setErrors([DOCUMENT_CREATION_ERRORS.NO_VALID_FILE]);
+      }
+    },
+    onSubmitJson: () => {
+      const { textInput } = values;
+
+      const MAX_UPLOAD_BYTES = 50 * 1000000; // 50 MB
+      if (Buffer.byteLength(textInput) > MAX_UPLOAD_BYTES) {
+        actions.setWarnings([DOCUMENT_CREATION_WARNINGS.LARGE_FILE]);
+      }
+
+      let documents;
+      try {
+        documents = JSON.parse(textInput);
+      } catch (error) {
+        return actions.setErrors([error.message]);
+      }
+
+      if (Array.isArray(documents)) {
+        actions.uploadDocuments({ documents });
+      } else if (isPlainObject(documents)) {
+        actions.uploadDocuments({ documents: [documents] });
+      } else {
+        actions.setErrors([DOCUMENT_CREATION_ERRORS.NOT_VALID]);
+      }
+    },
+    uploadDocuments: async ({ documents }) => {
+      const { http } = HttpLogic.values;
+      const { engineName } = EngineLogic.values;
+
+      const CHUNK_SIZE = 100;
+      const MAX_EXAMPLES = 5;
+
+      const promises = chunk(documents, CHUNK_SIZE).map((documentsChunk) => {
+        const body = JSON.stringify({ documents: documentsChunk });
+        return http.post(`/api/app_search/engines/${engineName}/documents`, { body });
+      });
+
+      try {
+        const responses = await Promise.all(promises);
+        const summary: DocumentCreationSummary = {
+          errors: [],
+          validDocuments: { total: 0, examples: [] },
+          invalidDocuments: { total: 0, examples: [] },
+          newSchemaFields: [],
+        };
+        responses.map((response) => {
+          if (response.errors?.length > 0) {
+            summary.errors = uniq([...summary.errors, ...response.errors]);
+            return;
+          }
+          summary.validDocuments.total += response.validDocuments.total;
+          summary.invalidDocuments.total += response.invalidDocuments.total;
+          summary.validDocuments.examples = [
+            ...summary.validDocuments.examples,
+            ...response.validDocuments.examples,
+          ].slice(0, MAX_EXAMPLES);
+          summary.invalidDocuments.examples = [
+            ...summary.invalidDocuments.examples,
+            ...response.invalidDocuments.examples,
+          ].slice(0, MAX_EXAMPLES);
+          summary.newSchemaFields = uniq([...summary.newSchemaFields, ...response.newSchemaFields]);
```

Comment:
This is all code I left as-is from the ent-search repo - in theory this could be a `.reduce` instead of a `.map`, but I actually think it's fine/more readable as is? 🤔 Any thoughts/preferences?

Reply:
It's fine as is. a `forEach` would have been more appropriate than a `map` though.

Reply:
oh snap, good catch!

Reply:
https://github.com/elastic/kibana/pull/86822/commits/f306732a30b4f1ffd03d38e023620fc5318ce6e4